File: q10-valid-parentheses/q10-sol.js
// LeetCode 20
// https://leetcode.com/problems/valid-parentheses/

const parens = {
    '{': '}',
    '[': ']',
    '(': ')'
}

const isValid = function(s) {
    
    if(!s) return true;
    
    let stack = [];
    for (let i = 0; i < s.length; i++) {
        if(parens[s[i]]) {
            stack.push(s[i]);
        } else {  
            if (s[i] !== parens[stack.pop()]) 
                return false;
        }
    }
    return stack.length === 0;
};

// time: O(n)
// space: O(n)

File: ztm-ds-algo/4-stacks-and-queues/q10-valid-parentheses/q10-sol.py
# LeetCode 20
# https://leetcode.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        charMap = {'(': ')', '{': '}', '[': ']'}
        for char in s:
            if char in charMap:
                stack.append(char)
            else:
                if len(stack) == 0 or char != charMap[stack.pop()]:
                    return False
        return len(stack) == 0

# time: O(n)
# space: O(n)